src/xopen/__init__.py

```diff
@@ -308,20 +311,30 @@ def __repr__(self):
             self._threads,
         )
 
-    def close(self) -> None:
-        if self.closed:
-            return
-        self.closed = True
+    def _terminate(self):
```

Comment:
Writing `_terminate` this way leads to some unnecessary code duplication down below when you need to override it in the `PipedPBzip2Writer` class. I also wouldn’t expect a function that is named `_terminate` to return an allowed error code and an allowed error message.

Perhaps you don’t need to add a `_terminate` method at all. You could just add class-level attributes to the `PipedCompressionWriter` like this:
```python
class PipedCompressionWriter(...):
    # This exit code is not interpreted as an error when terminating the process
    allowed_exit_code: Optional[int] = None
    # If this message is printed on stderr on terminating the process,
    # it is not interpreted as an error
    allowed_exit_message: Optional[str] = None
```
Then you set it appropriately in the subclasses and use it in `close()`.

Reply:
@marcelm  Adjusted. I was wondering if we should make PipedCompressionWriter abstract and use abstract properties for this. This would force future expansions (e.g. parallel xz?) to check what the correct values for allowed_exit_code and allowed_exit_message should be and not use the defaults of `None`

Reply:
This would be a nice improvement. I’ll open an issue so we won’t forget (let’s do that in a separate PR).